File: leetcode/447_number-of-boomerangs/447.py
class Solution(object):
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        count, n = 0, len(points)
        for i in range(n):
            distMap = {}
            for j in range(n):
                dist = (points[i][0] - points[j][0]) ** 2 + (points[i][1] - points[j][1]) ** 2
                if dist not in distMap: distMap[dist] = 0
                count += distMap[dist] * 2
                distMap[dist] += 1
        return count


if __name__ == '__main__':
    s = Solution()
    print (s.numberOfBoomerangs([[0,0],[1,0],[2,0]]))

File: leetcode/447_number-of-boomerangs/447.cpp
#include <algorithm>
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
    int numberOfBoomerangs(vector<pair<int, int>>& points) {
        int count = 0;
        for (vector<pair<int, int>>::const_iterator it = points.begin();
             it != points.end(); it++) {
            map<int, int> distMap;
            for (vector<pair<int, int>>::const_iterator ij = points.begin();
                ij != points.end(); ij++) {
                if (it == ij) continue;
                int dist = (it->first - ij->first) * (it->first - ij->first)
                         + (it->second - ij->second) * (it->second - ij->second);
                
                if (distMap.find(dist) == distMap.end()) {
                    distMap[dist] = 0;
                }
                count += distMap[dist] * 2;
                distMap[dist]++;
            }
        }
        return count;
    }
};

